src/rpc/rawtransaction.cpp

```diff
@@ -328,6 +328,10 @@ static UniValue verifytxoutproof(const JSONRPCRequest& request)
         throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Block not found in chain");
     }
 
+    if (pindex->nTx != merkleBlock.txn.GetNumTransactions()) {
```

Comment:
Is it right to throw an RPC error here? This is more similar to being given an invalid proof (as opposed to an incorrectly formatted/unusable one), in which case Null is returned currently.

Reply:
deja vu: https://github.com/bitcoin/bitcoin/pull/11120#discussion_r174324384

Reply:
Yes, I feel there is a big difference between not having the block (which means being unable to validate) and being given an intentionally fraudulent proof.

Requiring software to deal and match RPC errors instead of being covered by dealing with the existing "invalid proof" case sounds much nicer.

I also think you need to check for `nTx = 0` by the way (as it may mean a block for which we only have headers etc), and treat that as not having the block.

Reply:
@sipa Is it well-defined that a headers-only blockindex will have nTx==0? The comments say nTx simply is unreliable in that case...